### PR TITLE
Favourites tab

### PR DESCRIPTION
**Pull request by ffsiren**
This is a "bug fix" version of favourites a.k.a just to get something visible for the end users to comment on

https://digitransit.atlassian.net/browse/DT-173
https://digitransit.atlassian.net/browse/DT-172

### CODE REVIEW

File: app/component/favourites/favourite-route-row.cjsx [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+Icon                  = require '../icon/icon'
+Link                  = require 'react-router/lib/Link'
+classNames            = require 'classnames'
+FavouriteRouteActions = require '../../action/favourite-routes-action'
+RouteNumber           = require '../departure/route-number'
+
+class FavouriteRouteRow extends React.Component
+
+  proptypes: {
+    route: React.PropTypes.func.isRequired
```
Comment (hannesj) [inline app/component/favourites/favourite-route-row.cjsx]: Shouldn't this be React.PropTypes.object.isRequired?

File: app/component/favourites/favourite-route-row.cjsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+  addFavouriteRoute: (e) =>
+    @context.executeAction FavouriteRouteActions.addFavouriteRoute, @props.route.gtfsId
+
+  render: =>
+    className = @props.route.type.toLowerCase()
+    <div key={@props.route.gtfsId} className="small-12 medium-6 large-4 columns">
+        <div className="card padding-small cursor-pointer">
+          <p className="route-detail-text no-padding no-margin">
+            <Link to="#{process.env.ROOT_PATH}linjat/#{@props.route.gtfsId}:0:01" className="no-decoration">
```
Comment (hannesj) [inline app/component/favourites/favourite-route-row.cjsx]: We should probably get patterns {gtfsId} from relay, as direction can be empty.

File: app/component/favourites/favourite-stop-card-list-container.cjsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+class FavouriteStopCardListContainer extends React.Component
+
+  proptypes: {
+    stops: React.PropTypes.any.isRequired
+  }
+
+  getStopCards: =>
+    stopCards = []
+    for stop in @props.stops.stops
+      stopCards.push <StopCardContainer key={stop.gtfsId} stop={{stop:stop}} departures=DEPARTURES_COUNT />
```
Comment (hannesj) [inline app/component/favourites/favourite-stop-card-list-container.cjsx]: This also has unnecessary curly braces. Also, there should be a space added after the colon, probably would be a good idea to run coffeelint, with the config from the root folder.

File: app/component/favourites/favourite-stop-card-list-container.cjsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    stopCards
+
+  render: =>
+    <StopCardList className="row" addStops=false>
+    	{@getStopCards()}
```
Comment (hannesj) [inline app/component/favourites/favourite-stop-card-list-container.cjsx]: We are using only spaces, no tabs in the project.

File: app/component/favourites/favourites-panel.cjsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      Component={FavouriteRouteListContainer}
+      route={new queries.FavouriteRouteRowRoute({
+        ids: ids
+        })}
+      renderLoading={-> <div className="spinner-loader"/>}
+      }
+    />
+
+  componentDidMount: ->
+    @context.getStore('FavouriteRoutesStore').addChangeListener @onChange
```
Comment (hannesj) [inline app/component/favourites/favourites-panel.cjsx]: Shouldn't we also listen to FavouriteStopsStore

File: app/component/favourites/favourites-panel.cjsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  componentWillUnmount: ->
+    @context.getStore('FavouriteRoutesStore').removeChangeListener @onChange
+
+  onChange: (id) =>
+    @forceUpdate()
+
+  getFavourites: (stops, routes) =>
```
Comment (hannesj) [inline app/component/favourites/favourites-panel.cjsx]: Is this really needed, couldn't we just have 

  render: ->
    FavouriteStopsStore = @context.getStore 'FavouriteStopsStore'
    FavouriteRoutesStore = @context.getStore 'FavouriteRoutesStore'

    if stops.concat(routes).length==0
      <NoFavouritesPanel/>
    else
      <div>
        {@getFavouriteStopContainer FavouriteStopsStore.getStops()}
        {@getFavouriteRouteListContainer FavouriteRoutesStore.getRoutes()}
      </div>

File: app/component/favourites/no-favourites-panel.cjsx [hunk 7]
```diff
@@ -0,0 +1,12 @@
+React          = require 'react'
+
+class NoFavouritesPanel extends React.Component
+
+  render: ->
+    <div className="gray text-center">
+      <p>
+        Et ole vielä tallentanut yhtään suosikkia, joten emme voi näyttää niitä.
```
Comment (hannesj) [inline app/component/favourites/no-favourites-panel.cjsx]: TODO: localization

File: app/component/stop-cards/stop-card-list.cjsx [hunk 8]
```diff
@@ -12,13 +12,14 @@ class StopCardList extends React.Component
...
         </MasonryComponent>
       </div>
-      <div className="row">
-        <div className="small-10 small-offset-1 medium-6 medium-offset-3 columns">
-          <button className="show-more cursor-pointer" onClick=@props.addStops>
-            <FormattedMessage id="show-more" defaultMessage="Show more" />
-          </button>
-        </div>
-      </div>
+      {if @props.stops && @props.stops.stopsByRadius.pageInfo.hasNextPage
```
Comment (hannesj) [inline app/component/stop-cards/stop-card-list.cjsx]: AddStops is used to increase the max radius, after the stop list has been exhausted for nearest stops, so this should probably have also a toggle in props.

File: app/queries.js [hunk 9]
```diff
@@ -135,6 +135,35 @@ var StopListContainerFragments = {
   `,
 };
 
+class FavouriteStopListContainerRoute extends Relay.Route {
+  static queries = {
+      stops: (Component, variables) => Relay.QL`
+      query {
+        viewer {
```
Comment (hannesj) [inline app/queries.js]: You don't need a viewer, but can use stops(ids: $ids) directly here.

**Comment by ffsiren** [comment]
Thanks Hannes for the review. I believe I have now addressed all the commented points. Ready to go in?

File: app/queries.js [hunk 10]
```diff
@@ -338,6 +365,39 @@ var RouteMarkerPopupFragments = {
   `,
 }
 
+class FavouriteRouteRowRoute extends Relay.Route {
```
Comment (hannesj) [inline app/queries.js]: You could do the same improvements as in 4c1812fc9b1206ec63f981300ae4058ebf4a3142 for this route also

**Comment by ffsiren** [comment]
The route Route does not use viewer anymore.

**Pull request merged**